fix check of bounds on parameters sorting

fdplacido opened the pull request:
BEGINRELEASENOTES
- Fix checks on second parameter when sorting them.

ENDRELEASENOTES

andresailer commented on k4MarlinWrapper/src/components/EDM4hep2Lcio.cpp:
```diff
@@ -1162,14 +1162,16 @@ void EDM4hep2LcioTool::optimizeOrderParams()
-          std::swap(m_edm2lcio_params[track_index], m_edm2lcio_params[trackerhit_index]);
-          std::swap(m_edm2lcio_params[track_index+1], m_edm2lcio_params[trackerhit_index+1]);
-          std::swap(m_edm2lcio_params[track_index+2], m_edm2lcio_params[trackerhit_index+2]);
+        if (trackerhit_found_it != m_edm2lcio_params.end()) {
```
this way less indentation changes are needed below.
```suggestion
        if (trackerhit_found_it == m_edm2lcio_params.end()) {
          continue;
        }
```

vvolkl commented:
Could you add a CMake option  `USE_EXTERNAL_CATCH2` like in k4simdelphes

https://github.com/key4hep/k4SimDelphes/blob/main/tests/unittests/CMakeLists.txt#L2
https://github.com/key4hep/k4SimDelphes/blob/main/CMakeLists.txt#L111

? This will let us build a static library for catch2 once  instead of rebuilding it for every repository.

fdplacido commented:
> Could you add a CMake option `USE_EXTERNAL_CATCH2` like in k4simdelphes
> 
> https://github.com/key4hep/k4SimDelphes/blob/main/tests/unittests/CMakeLists.txt#L2
> https://github.com/key4hep/k4SimDelphes/blob/main/CMakeLists.txt#L111
> 
> ? This will let us build a static library for catch2 once instead of rebuilding it for every repository.

I implemented it, thanks!

andresailer commented on k4MarlinWrapper/k4MarlinWrapper/util/EDM4hep2LcioUtils.h:
```diff
@@ -0,0 +1,60 @@
+    // if target before dependency swap all related elements. Multiply indexes by 3 to get correct index from view.
+    auto swap_if_before = [&params_view, &params](const std::string& target, const std::string& dependency)
+    {
+      auto target_found_it = std::find(params_view.begin(), params_view.end(), target);
```
you need to find all the targets, as there can be more than one

fdplacido replied:
I changed the algorithm to take into account multiple repeated types, so it now searches for all targets, dependencies, etc.
I added a unit test with that case.

fdplacido commented:
I changed the algorithm to take into account multiple repeated types, so it now searches for all targets, dependencies, etc.
I added a unit test with that case.

fdplacido commented on k4MarlinWrapper/k4MarlinWrapper/util/EDM4hep2LcioUtils.h:
```diff
@@ -0,0 +1,60 @@
+          auto dependency_index = 3 * std::distance(params_view.begin(), dependency_found_it);
+
+          if (target_index < dependency_index) {
+            std::swap(params[target_index], params[dependency_index]);
```
I didn't know about this. It needs to work with iterators which in this case would add more lines. I think for 3 cases its still simpler this way.

andresailer commented on k4MarlinWrapper/k4MarlinWrapper/util/EDM4hep2LcioUtils.h:
```diff
@@ -0,0 +1,72 @@
+    // if target before dependency swap all related elements. Multiply indexes by 3 to get correct index from view.
+    auto swap_if_before = [&params_view, &params](const std::string& target, const std::string& dependency)
+    {
+      const int maxidx = 5;
```
unused?

andresailer commented on k4MarlinWrapper/k4MarlinWrapper/util/EDM4hep2LcioUtils.h:
```diff
@@ -0,0 +1,72 @@
+    auto swap_if_before = [&params_view, &params](const std::string& target, const std::string& dependency)
+    {
+      const int maxidx = 5;
+      int gidx = 0;
```
unused?

andresailer commented on k4MarlinWrapper/k4MarlinWrapper/util/EDM4hep2LcioUtils.h:
```diff
@@ -0,0 +1,72 @@
+    };
+
+    // A depends on B converted before to not need FillMissingCollections()
+    for (int i=0; i < 2; ++i) {
```
why this loop?

fdplacido replied:
With just 1 loop, some collections will go "up" again in the list and some targets will stay before the dependency. Going over them again fixes that.

andresailer commented on k4MarlinWrapper/k4MarlinWrapper/util/EDM4hep2LcioUtils.h:
```diff
@@ -0,0 +1,72 @@
+          // Look for instances of dependency
+          for (int d_idx=0; d_idx<params_view.size(); ++d_idx) {
+            if (params_view[d_idx] == dependency) {
+              // If target is before the depenency swap them
```
```suggestion
              // If target is before the dependency swap them
```

andresailer commented on k4MarlinWrapper/k4MarlinWrapper/util/EDM4hep2LcioUtils.h:
```diff
@@ -0,0 +1,71 @@
+  T& params)
+{
+
+  if (params.size() % 3 == 0) {
```
Do you really need this check? Is this not checked much earlier already and then issuing a fatal error?

fdplacido replied:
It is indeed, this was an early attempt to make the method "more independent" from the only place that it can be really called. But I think that's a branch we can remove.

andresailer replied:
yes, either remove this branch or throw an exception if this isn't fulfilled?

fdplacido replied:
I removed to not keep two checks.
